### PR TITLE
feat: add tool calling for cohere

### PR DESCRIPTION
**Pull request by Anirudh31415926535**
# Description

Adds function calling for Cohere integration in llama-index.

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py [hunk 1]
```diff
@@ -131,13 +140,169 @@ def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
...
+        if not tool_calls:
+            tool_calls = []
+
+        if len(tool_calls) < 1:
+            if error_on_no_tool_call:
+                raise ValueError(
+                    f"Expected at least one tool call, but got {len(tool_calls)} tool calls."
+                )
+            else:
+                return []
```
Comment (shaandes) [inline llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py:184]: could we make this the following
```suggestion
        if not tool_calls and error_on_no_tool_call:
                raise ValueError(
                    f"Expected at least one tool call, but got {len(tool_calls)} tool calls."
                )
      else:
          return []
```

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py [hunk 2]
```diff
@@ -131,13 +140,169 @@ def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
...
+
+        tool_selections = []
+        for tool_call in tool_calls:
+            if not isinstance(tool_call, ToolCall):
+                raise ValueError("Invalid tool_call object")
+            # argument_dict = json.loads(tool_call.parameters)
+            tool_selections.append(
+                ToolSelection(
+                    tool_id="TESTESTSE",
```
Comment (shaandes) [inline llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py:193]: why not pass the actual id?

Comment (Anirudh31415926535) [reply llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py:193]: Oh shit this was a test id... I'll update haha thanks for catching!

File: pyproject.toml [hunk 3]
```diff
@@ -47,7 +47,7 @@ repository = "https://github.com/run-llama/llama_index"
 version = "0.10.59"
 
 [tool.poetry.dependencies]
-python = ">=3.8.1,<4.0"
+python = ">=3.9,<4.0"
```
Comment (logan-markewich) [inline pyproject.toml:50]: uhhh we don't want to do this on the root package

File: pyproject.toml [hunk 4]
```diff
@@ -18,7 +18,7 @@ explicit_package_bases = true
 ignore_missing_imports = true
 mypy_path = "llama_index"
 namespace_packages = true
-python_version = "3.8"
+python_version = "3.9"
```
Comment (logan-markewich) [inline pyproject.toml:21]: uhhh we don't want to do this on the root package

File: pyproject.toml (hunk 3 above)
Comment (logan-markewich) [reply pyproject.toml]: reverted

File: pyproject.toml (hunk 4 above)
Comment (logan-markewich) [reply pyproject.toml]: reverted

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py (hunk 1 above)
Comment (Anirudh31415926535) [reply llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py]: Updated the logic here

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py [hunk 5]
```diff
@@ -131,13 +141,164 @@ def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
...
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> Dict[str, Any]:
+        """Prepare the chat with tools."""
+        chat_history = chat_history or []
+
+        if isinstance(user_msg, str):
+            user_msg = ChatMessage(role=MessageRole.USER, content=user_msg)
+            chat_history.append(user_msg)
```
Comment (harry-cohere) [inline llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py:158]: Is this correct?
```suggestion
        chat_history.append(user_msg)
```

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py [hunk 6]
```diff
@@ -131,13 +141,164 @@ def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
...
+        if len(tool_calls) < 1 and error_on_no_tool_call:
+            raise ValueError(
+                f"Expected at least one tool call, but got {len(tool_calls)} tool calls."
+            )
+
+        tool_selections = []
+        for tool_call in tool_calls:
+            if not isinstance(tool_call, ToolCall):
+                raise ValueError("Invalid tool_call object")
+            # argument_dict = json.loads(tool_call.parameters)
```
Comment (harry-cohere) [inline llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/base.py:186]: Might not need this anymore
```suggestion
```

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/utils.py [hunk 7]
```diff
@@ -312,3 +370,161 @@ def document_message_to_cohere_document(message: DocumentMessage) -> Dict:
...
+
+
+def _remove_signature_from_tool_description(name: str, description: str) -> str:
+    """
+    Removes the `{name}{signature} - ` prefix and Args: section from tool description.
+    The signature is usually present for tools created with the @tool decorator,
+    whereas the Args: section may be present in function doc blocks.
+    """
+    description = re.sub(rf"^{name}\(.*?\) -(?:> \w+? -)? ", "", description)
+    return re.sub(r"(?s)(?:\n?\n\s*?)?Args:.*$", "", description)
```
Comment (harry-cohere) [inline llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/utils.py:382]: Is this needed? The comment might need tweaking, too

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/utils.py [hunk 8]
```diff
@@ -312,3 +370,161 @@ def document_message_to_cohere_document(message: DocumentMessage) -> Dict:
...
+    tools: Sequence[Union[Dict[str, Any], Type[BaseModel], Callable, BaseTool]],
+) -> List[Dict[str, Any]]:
+    return [_convert_to_cohere_tool(tool) for tool in tools]
+
+
+def _convert_to_cohere_tool(
+    tool: BaseTool,
+) -> Dict[str, Any]:
+    """
+    Convert a BaseTool instance, JSON schema dict, or BaseModel type to a Cohere tool.
```
Comment (harry-cohere) [inline llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/utils.py:395]:
```suggestion
    Convert a BaseTool instance to a Cohere tool.
```

**Review by harry-cohere — COMMENTED** [review]
Looks great!

File: llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/utils.py (hunk 7 above)
Comment (Anirudh31415926535) [reply llama-index-integrations/llms/llama-index-llms-cohere/llama_index/llms/cohere/utils.py:382]: Updated comment.

**Comment by logan-markewich** [comment]
@Anirudh31415926535 @harry-cohere is this good to go from your point of view?

**Pull request merged**